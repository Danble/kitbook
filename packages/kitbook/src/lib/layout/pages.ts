export type Folder = {
  name: string;
  url: string;
  depth: number;
  folders?: Folder[];
  pages?: Page[];
};

// create 1 page, show any stories/docs first, then default component view, then any variants
export type Page = {
  name: string;
  ext: string;
  path: string;
  url: string;
  // type?: 'story' | 'component' | 'page' | 'variants';
  // 1st check for story/docs (md/svx)
  storyModulePath?: string;
  // 2nd check for component or page (svelte)
  componentModulePath?: string;
  pageModulePath?: string;
  // 3rd check for variants (variants.ts)
  variantsModulePath?: string;
};

export type Modules = Record<string, () => Promise<{ [key: string]: any }>>;

function removeInitialDigitAndHyphens(string: string) {
  return string.replace(/^\d+/, '').replace(/-/g, ' ').trim();
}

export function parsePath(path: string) {
  if (path === '/README.md') return { ext: 'md', name: 'README', dir: '/' }

  const match = path.match(/^\/src\/(.*\/)(.+?)\.(.+)$/);
  if (!match) throw new Error(`${path} is not a module path that Kitbook can handle. Make sure your Kitbook Layout Load import meta glob starts with '/src/**'`);
  const [, dir, name, ext] = match;
  return { dir, name, ext };
}

export function parseModules(modules: Modules, root = '/kitbook'): Page[] {
  const paths = Object.keys(modules)
    .filter(p => {
      return !p.includes(`/src/routes${root}`) &&
        !p.includes('+layout.svelte')
    });

  if (!paths.length) return []

  const pages = paths.map((path) => {
    const { name, ext } = parsePath(path);
    const url = path.replace('src/', '').replace(`.${ext}`, '');
    return {
      path,
      name: removeInitialDigitAndHyphens(name),
      ext,
      url,
    };
  });
  return pages;
}

export function combineModulesIntoPages(uncombined: Page[]): Page[] {
  return uncombined;
}
// Will skip file if an unexpected additional . is found in file name, e.g. /src/A/Bar.foo.svelte

export function putPagesIntoFolders(pagesToOrganize: Page[]): Folder {
  const rootFolder: Folder = {
    name: '.',
    url: '/',
    depth: 0,
    folders: [],
    pages: [],
  };

  if (!pagesToOrganize?.length) return rootFolder;

  pagesToOrganize.forEach((page) => {
    let currentFolder = rootFolder;

    const path = page.url.split('/');
    path.forEach((folderName, index) => {
      const cleanedFolderName = removeInitialDigitAndHyphens(folderName);

      if (index === path.length - 1) {
        currentFolder.pages.push(page); // is a page
      } else if (folderName === '') {
        currentFolder = rootFolder;
      } else {
        let folder = currentFolder.folders.find((folder) => folder.name === cleanedFolderName);
        if (!folder) {
          folder = {
            name: cleanedFolderName,
            url: path
              .slice(0, index + 1)
              .join('/')
              .replace(/^\./, ''),
            depth: index,
            folders: [],
            pages: [],
          };
          currentFolder.folders.push(folder);
        }
        currentFolder = folder;
      }
    });
  });
  return rootFolder;
}


