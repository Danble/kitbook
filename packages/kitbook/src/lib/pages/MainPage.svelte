<script lang="ts">
  import { page } from '$app/stores';
  import { compressToEncodedURIComponent as encode } from 'lz-string';
  import type { GroupedPage, GroupedPageMap, LoadedModules } from '../kitbook-types';
  import EditInGithub from '../components/EditInGithub.svelte';

  export let data: {
    pages?: GroupedPageMap;
    page?: GroupedPage;
    loadedModules?: LoadedModules;
    error?: string;
  } = { loadedModules: {} };

  $: pathWouldRecurseInfinitelyIfInSandbox = $page.url.pathname.startsWith(
    '/lib/routes/sandbox/[...file]/+'
  );
  $: doesNotHaveSvxOrVariants = !(data.loadedModules.svx || data.loadedModules.variants);
</script>

{#if data.error}
  {data.error}
{:else}
  {#if data.loadedModules.svx}
    <div class="mb-10">
      <svelte:component this={data.loadedModules.svx} />
    </div>
  {/if}

  {#if data.loadedModules.component}
    {#if data.loadedModules.variants}
      <div class="text-2xl">
        {data.page.name.startsWith('+') ? 'Page' : 'Component'} Variants
      </div>
      {#each data.loadedModules.variants as variant, index}
        <div class="not-prose border rounded mt-3">
          <div class="bg-gray-200 p-3 mb-2">
            <span class="font-semibold">
              {variant.name}
            </span>
            {#if variant.description}
              <div class="text-sm">
                {variant.description}
              </div>
            {/if}
          </div>
          {#if pathWouldRecurseInfinitelyIfInSandbox}
            <svelte:component this={data.loadedModules.component} {...variant.props || {}} />
          {:else}
            <iframe
              class="w-full h-full"
              title=""
              src="/sandbox/{$page.url.pathname}?props={encode(
                JSON.stringify(variant.props || {})
              )}&variantIdx={index}"
            />
          {/if}
        </div>
      {/each}
      <!-- {:else if !pathWouldRecurseInfinitelyIfInSandbox}
      <div class="not-prose border rounded mt-3">
        <div class="bg-gray-200 p-3 mb-2">
          <span class="font-semibold"> Default </span>
        </div>
        <iframe
          class="w-full h-full"
          title=""
          src="/sandbox/{$page.url.pathname}?props={encode(JSON.stringify({}))}"
        />
      </div> -->
    {/if}
  {/if}

  {#if doesNotHaveSvxOrVariants}
    <div class="mb-3 p-3 bg-gray-200 rounded">
      <b>Kitbook tip</b>: You have not created a Stories file ({data.page?.name}.svx/.md) nor a Variants file ({data
        .page?.name}.variants.ts) file. In the future Kitbook will try to
      automatically supply default props, but until then you must manually scaffold a page for this
      component by creating either a Stories or Variants file.
    </div>
  {/if}

  <EditInGithub path={data?.page?.path} />
{/if}
