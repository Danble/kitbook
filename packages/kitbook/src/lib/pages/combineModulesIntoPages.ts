import type { Page, PageMap } from "$lib/kitbook-types";
import { testModules } from "./testModules";
import { parseModules } from "./parseModules";

/** allow _page/_layout to be merged with +page/+layout */
function convertUnderscorePrefixToPlus(s: string): string {
  return s.replace('_page', '+page').replace('_layout', '+layout')
}

export function combineModulesIntoPages(uncombined: Page[]): PageMap {
  const combined: PageMap = {};

  for (const page of uncombined) {
    const url = convertUnderscorePrefixToPlus(page.url);
    if (!combined[url])
      combined[url] = { name: page.name, url: page.url }

    // Will skip files not matching expected extensions, e.g. /src/A/Bar.foo.svelte
    if (['md', 'svx'].includes(page.ext)) {
      combined[url].svxModulePath = page.path
    } else if ((page.name === '+page') || (page.name === '+layout')) {
      combined[url].pageModulePath = page.path // layouts are just pages with slot inheritance super-powers
    } else if (page.ext === 'svelte') {
      combined[url].componentModulePath = page.path
    } else if (page.ext === 'variants.ts') {
      combined[url].variantsModulePath = page.path
    }
  }

  return combined;
}

if (import.meta.vitest) {
  test('combineModulesIntoPages properly combines Pages', () => {
    const pages = parseModules(testModules);
    expect(combineModulesIntoPages(pages)).toMatchInlineSnapshot(`
      {
        "/README": {
          "name": "README",
          "svxModulePath": "/README.md",
          "url": "/README",
        },
        "/docs/0-why-kitbook": {
          "name": "why kitbook",
          "svxModulePath": "/src/docs/0-why-kitbook.md",
          "url": "/docs/0-why-kitbook",
        },
        "/docs/1-get-started": {
          "name": "get started",
          "svxModulePath": "/src/docs/1-get-started.md",
          "url": "/docs/1-get-started",
        },
        "/docs/index": {
          "name": "index",
          "svxModulePath": "/src/docs/index.md",
          "url": "/docs/index",
        },
        "/docs/my-notes/0-unocss": {
          "name": "unocss",
          "svxModulePath": "/src/docs/my-notes/0-unocss.md",
          "url": "/docs/my-notes/0-unocss",
        },
        "/docs/my-notes/1-deploy-to-vercel": {
          "name": "deploy to vercel",
          "svxModulePath": "/src/docs/my-notes/1-deploy-to-vercel.md",
          "url": "/docs/my-notes/1-deploy-to-vercel",
        },
        "/lib/A": {
          "componentModulePath": "/src/lib/A.svelte",
          "name": "A",
          "url": "/lib/A",
        },
        "/lib/B": {
          "componentModulePath": "/src/lib/B.svelte",
          "name": "B",
          "svxModulePath": "/src/lib/B.svx",
          "url": "/lib/B",
        },
        "/lib/E": {
          "name": "E",
          "svxModulePath": "/src/lib/E.svx",
          "url": "/lib/E",
        },
        "/lib/a/C": {
          "componentModulePath": "/src/lib/a/C.svelte",
          "name": "C",
          "url": "/lib/a/C",
          "variantsModulePath": "/src/lib/a/C.variants.ts",
        },
        "/lib/a/D": {
          "componentModulePath": "/src/lib/a/D.svelte",
          "name": "D",
          "svxModulePath": "/src/lib/a/D.svx",
          "url": "/lib/a/D",
          "variantsModulePath": "/src/lib/a/D.variants.ts",
        },
        "/routes/+layout": {
          "name": "+layout",
          "pageModulePath": "/src/routes/+layout.svelte",
          "url": "/routes/+layout",
        },
        "/routes/+page": {
          "name": "+page",
          "pageModulePath": "/src/routes/+page.svelte",
          "url": "/routes/+page",
        },
        "/routes/a/+page": {
          "name": "+page",
          "pageModulePath": "/src/routes/a/+page.svelte",
          "svxModulePath": "/src/routes/a/_page.svx",
          "url": "/routes/a/+page",
        },
        "/routes/b/+page": {
          "name": "+page",
          "pageModulePath": "/src/routes/b/+page.svelte",
          "url": "/routes/b/+page",
          "variantsModulePath": "/src/routes/b/_page.variants.ts",
        },
        "/routes/c/+page": {
          "name": "+page",
          "pageModulePath": "/src/routes/c/+page.svelte",
          "svxModulePath": "/src/routes/c/_page.svx",
          "url": "/routes/c/+page",
          "variantsModulePath": "/src/routes/c/_page.variants.ts",
        },
      }
    `);
  });
}