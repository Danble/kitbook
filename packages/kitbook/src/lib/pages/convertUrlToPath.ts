export function convertUrlToModulePath(url: string, extension: 'md' | 'svx' | 'svelte' = 'svelte'): string {
  let path = extension === 'svelte' ? url : url.replace('/+page', '/_page');
  return `/src${path}.${extension}`;
}