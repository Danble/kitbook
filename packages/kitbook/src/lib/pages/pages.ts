import type { Folder, Page, Modules } from "kitbook";

type PageMap = Record<string, Page>;

function removeInitialDigitAndHyphens(string: string) {
  return string.replace(/^\d+/, '').replace(/-/g, ' ').trim();
}
/** allow _page/_layout to be merged with +page/+layout */
function convertUnderscorePrefixToPlus(s: string): string {
  return s.replace('_page', '+page').replace('_layout', '+layout')
}

export function parsePath(path: string) {
  if (path === '/README.md') return { ext: 'md', name: 'README', dir: '/' }

  const match = path.match(/^\/src\/(.*\/)(.+?)\.(.+)$/);
  if (!match) throw new Error(`${path} is not a module path that Kitbook can handle. Make sure your Kitbook Layout Load import meta glob starts with '/src/**'`);
  const [, dir, name, ext] = match;
  return { dir, name, ext };
}

export function parseModules(modules: Modules): Page[] {
  const paths = Object.keys(modules)
    .filter(p => {
      return !p.includes(`/src/kitbook`) &&
        !p.includes('+layout.svelte')
    });

  if (!paths.length) return []

  const pages = paths.map((path) => {
    const { name, ext } = parsePath(path);
    const url = path.replace('src/', '').replace(`.${ext}`, '');
    return {
      path,
      name: removeInitialDigitAndHyphens(name),
      ext,
      url,
    };
  });
  return pages;
}

export function combineModulesIntoPages(uncombined: Page[]): PageMap {
  const combined: PageMap = {};

  for (const page of uncombined) {
    const url = convertUnderscorePrefixToPlus(page.url);
    if (!combined[url])
      combined[url] = { name: page.name, url: page.url }

    // Will skip files not matching expected extensions, e.g. /src/A/Bar.foo.svelte
    if (['md', 'svx'].includes(page.ext)) {
      combined[url].svxModulePath = page.path
    } else if (page.name === '+page' || page.name === '+layout') {
      combined[url].pageModulePath = page.path // layouts are just pages with slot inheritance super-powers
    } else if (page.ext === 'svelte') {
      combined[url].componentModulePath = page.path
    } else if (page.ext === 'variants.ts') {
      combined[url].variantsModulePath = page.path
    }
  }

  return combined;
}

export function putPagesIntoFolders(combinedPages: PageMap): Folder {
  const rootFolder: Folder = {
    name: '.',
    url: '/',
    depth: 0,
    folders: [],
    pages: [],
  };

  const pagesToOrganize = Object.values(combinedPages)
  if (!pagesToOrganize?.length) return rootFolder;

  pagesToOrganize.forEach((page) => {
    let currentFolder = rootFolder;

    const path = page.url.split('/');
    path.forEach((folderName, index) => {
      const cleanedFolderName = removeInitialDigitAndHyphens(folderName);

      if (index === path.length - 1) {
        currentFolder.pages.push(page); // is a page
      } else if (folderName === '') {
        currentFolder = rootFolder;
      } else {
        let folder = currentFolder.folders.find((folder) => folder.name === cleanedFolderName);
        if (!folder) {
          folder = {
            name: cleanedFolderName,
            url: path
              .slice(0, index + 1)
              .join('/')
              .replace(/^\./, ''),
            depth: index,
            folders: [],
            pages: [],
          };
          currentFolder.folders.push(folder);
        }
        currentFolder = folder;
      }
    });
  });
  return rootFolder;
}


