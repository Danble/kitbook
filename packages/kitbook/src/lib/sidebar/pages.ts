export type Folder = {
  name: string;
  url: string;
  depth: number;
  folders?: Folder[];
  pages?: Page[];
};

export type Page = {
  name: string;
  ext: string;
  path: string;
  url: string;
};

export type Modules = Record<string, () => Promise<{ [key: string]: any }>>;

function removeInitialDigitAndHyphens(string: string) {
  return string.replace(/^\d+/, '').replace(/-/g, ' ').trim();
}

export function parsePath(path: string) {
  if (path === '/README.md') return { ext: 'md', name: 'README', dir: '/' }

  const match = path.match(/^\/src\/(.*\/)(.+?)\.(.+)$/);
  if (!match) throw new Error(`${path} is not a module path that Kitbook can handle. Make sure your Kitbook Layout Load import meta glob starts with '/src/**'`);
  const [, dir, name, ext] = match;
  return { dir, name, ext };
}

export function parseModules(modules: Modules, root = '/kitbook'): Page[] {
  const paths = Object.keys(modules)
    .filter(p => {
      return !p.includes(`/src/routes${root}`) &&
        !p.includes('+layout.svelte')
    });

  if (!paths.length) return []

  const pages = paths.map((path) => {
    const { dir, name, ext } = parsePath(path);
    return {
      path: path.replace('/src/', ''),
      name: removeInitialDigitAndHyphens(name),
      ext,
      url: dir,
    };
  });
  return pages;
}

export function putPagesIntoFolders(pagesToOrganize: Page[]): Folder {
  const rootFolder: Folder = {
    name: '.',
    url: '/',
    depth: 0,
    folders: [],
    pages: [],
  };

  if (!pagesToOrganize?.length) return rootFolder;

  pagesToOrganize.forEach((page) => {
    let currentFolder = rootFolder;

    const path = page.url.split('/');
    path.forEach((folderName, index) => {
      const cleanedFolderName = removeInitialDigitAndHyphens(folderName);

      if (index === path.length - 1) {
        currentFolder.pages.push(page); // is a page
      } else if (folderName === '') {
        currentFolder = rootFolder;
      } else {
        let folder = currentFolder.folders.find((folder) => folder.name === cleanedFolderName);
        if (!folder) {
          folder = {
            name: cleanedFolderName,
            url: path
              .slice(0, index + 1)
              .join('/')
              .replace(/^\./, ''),
            depth: index + 1,
            folders: [],
            pages: [],
          };
          currentFolder.folders.push(folder);
        }
        currentFolder = folder;
      }
    });
  });
  return rootFolder;
}

// export function findActivePage(pages: Page[], urlPathname: string): Page {
//   const activePage = pages.find((page) => {
//     const regex = new RegExp(`${page.url}$`);
//     return regex.test(urlPathname);
//   });
//   if (activePage) return activePage;
//   return pages.find((page) => page.url === '/');
// }
