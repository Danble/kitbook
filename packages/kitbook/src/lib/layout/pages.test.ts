import { parseModules, putPagesIntoFolders, parsePath } from "./pages";

const modules = {
  '/README.md': () => Promise.resolve({}),

  // markdown docs
  '/src/docs/0-why-kitbook.md': () => Promise.resolve({}),
  '/src/docs/1-get-started.md': () => Promise.resolve({}),
  '/src/docs/index.md': () => Promise.resolve({}),
  '/src/docs/my-notes/0-unocss.md': () => Promise.resolve({}),
  '/src/docs/my-notes/1-deploy-to-vercel.md': () => Promise.resolve({}),
  
  // components
  '/src/lib/A.svelte': () => Promise.resolve({}), // by itself
  
  '/src/lib/B.svelte': () => Promise.resolve({}), // with svx
  '/src/lib/B.svx': () => Promise.resolve({}),
  
  '/src/lib/a/C.svelte': () => Promise.resolve({}), // with variants
  '/src/lib/a/C.variants.ts': () => Promise.resolve({}),
  
  '/src/lib/a/D.svelte': () => Promise.resolve({}), // with svx and variants
  '/src/lib/a/D.svx': () => Promise.resolve({}),
  '/src/lib/a/D.variants.ts': () => Promise.resolve({}),
  
  '/src/lib/E.svx': () => Promise.resolve({}), // svx by itself (e.g. display combinations of components)
  
  // pages
  '/src/routes/+page.svelte': () => Promise.resolve({}), // by itself
  
  '/src/routes/a/+page.svelte': () => Promise.resolve({}), // with svx
  '/src/routes/a/_page.svx': () => Promise.resolve({}),

  '/src/routes/b/+page.svelte': () => Promise.resolve({}), // with variants
  '/src/routes/b/_page.variants.ts': () => Promise.resolve({}),
  
  '/src/routes/c/+page.svelte': () => Promise.resolve({}), // with svx and variants
  '/src/routes/c/_page.svx': () => Promise.resolve({}),
  '/src/routes/c/_page.variants.ts': () => Promise.resolve({}),
  
  // ignore layout files and kitbook route files
  '/src/routes/+layout.svelte': () => Promise.resolve({}), 
  '/src/routes/kitbook/[...file]/+page.svelte': () => Promise.resolve({}),
  '/src/routes/kitbook/sandbox/[...id]/+page@.svelte': () => Promise.resolve({}),
  
  // unrecognized extensions will be ignored in combineModulesIntoPages()
  '/src/lib/A.foo.svelte': () => Promise.resolve({}),
}

test('putPagesIntoFolders organizes pages into proper folders', () => {
  const pages = parseModules(modules);
  expect(putPagesIntoFolders(pages)).toMatchSnapshot();
});

test('parseModules properly returns array of Page objects', () => {
  expect(parseModules(modules)).toMatchInlineSnapshot(`
    [
      {
        "ext": "md",
        "name": "README",
        "path": "/README.md",
        "url": "/README",
      },
      {
        "ext": "md",
        "name": "why kitbook",
        "path": "/src/docs/0-why-kitbook.md",
        "url": "/docs/0-why-kitbook",
      },
      {
        "ext": "md",
        "name": "get started",
        "path": "/src/docs/1-get-started.md",
        "url": "/docs/1-get-started",
      },
      {
        "ext": "md",
        "name": "index",
        "path": "/src/docs/index.md",
        "url": "/docs/index",
      },
      {
        "ext": "md",
        "name": "unocss",
        "path": "/src/docs/my-notes/0-unocss.md",
        "url": "/docs/my-notes/0-unocss",
      },
      {
        "ext": "md",
        "name": "deploy to vercel",
        "path": "/src/docs/my-notes/1-deploy-to-vercel.md",
        "url": "/docs/my-notes/1-deploy-to-vercel",
      },
      {
        "ext": "svelte",
        "name": "A",
        "path": "/src/lib/A.svelte",
        "url": "/lib/A",
      },
      {
        "ext": "svelte",
        "name": "B",
        "path": "/src/lib/B.svelte",
        "url": "/lib/B",
      },
      {
        "ext": "svx",
        "name": "B",
        "path": "/src/lib/B.svx",
        "url": "/lib/B",
      },
      {
        "ext": "svelte",
        "name": "C",
        "path": "/src/lib/a/C.svelte",
        "url": "/lib/a/C",
      },
      {
        "ext": "variants.ts",
        "name": "C",
        "path": "/src/lib/a/C.variants.ts",
        "url": "/lib/a/C",
      },
      {
        "ext": "svelte",
        "name": "D",
        "path": "/src/lib/a/D.svelte",
        "url": "/lib/a/D",
      },
      {
        "ext": "svx",
        "name": "D",
        "path": "/src/lib/a/D.svx",
        "url": "/lib/a/D",
      },
      {
        "ext": "variants.ts",
        "name": "D",
        "path": "/src/lib/a/D.variants.ts",
        "url": "/lib/a/D",
      },
      {
        "ext": "svx",
        "name": "E",
        "path": "/src/lib/E.svx",
        "url": "/lib/E",
      },
      {
        "ext": "svelte",
        "name": "+page",
        "path": "/src/routes/+page.svelte",
        "url": "/routes/+page",
      },
      {
        "ext": "svelte",
        "name": "+page",
        "path": "/src/routes/a/+page.svelte",
        "url": "/routes/a/+page",
      },
      {
        "ext": "svx",
        "name": "_page",
        "path": "/src/routes/a/_page.svx",
        "url": "/routes/a/_page",
      },
      {
        "ext": "svelte",
        "name": "+page",
        "path": "/src/routes/b/+page.svelte",
        "url": "/routes/b/+page",
      },
      {
        "ext": "variants.ts",
        "name": "_page",
        "path": "/src/routes/b/_page.variants.ts",
        "url": "/routes/b/_page",
      },
      {
        "ext": "svelte",
        "name": "+page",
        "path": "/src/routes/c/+page.svelte",
        "url": "/routes/c/+page",
      },
      {
        "ext": "svx",
        "name": "_page",
        "path": "/src/routes/c/_page.svx",
        "url": "/routes/c/_page",
      },
      {
        "ext": "variants.ts",
        "name": "_page",
        "path": "/src/routes/c/_page.variants.ts",
        "url": "/routes/c/_page",
      },
      {
        "ext": "foo.svelte",
        "name": "A",
        "path": "/src/lib/A.foo.svelte",
        "url": "/lib/A",
      },
    ]
  `);
});

test('parsePath parses path correctly', () => {
  expect(parsePath('/src/docs/0-why-kitbook.md')).toMatchInlineSnapshot(`
    {
      "dir": "docs/",
      "ext": "md",
      "name": "0-why-kitbook",
    }
  `);
  expect(parsePath('/src/docs/index.md')).toMatchInlineSnapshot(`
    {
      "dir": "docs/",
      "ext": "md",
      "name": "index",
    }
  `);
  expect(parsePath('/README.md')).toMatchInlineSnapshot(`
    {
      "dir": "/",
      "ext": "md",
      "name": "README",
    }
  `);
  expect(parsePath('/src/routes/a/+page.svelte')).toMatchInlineSnapshot(`
    {
      "dir": "routes/a/",
      "ext": "svelte",
      "name": "+page",
    }
  `);
  expect(parsePath('/src/routes/a/_page.variants.ts')).toMatchInlineSnapshot(`
    {
      "dir": "routes/a/",
      "ext": "variants.ts",
      "name": "_page",
    }
  `);
});

test.skip('skips and logs warning upon receiving unusable path', () => {
  expect(parsePath('+page.ts')).toThrow();
  // expect(parsePath('+page.ts')).toThrowErrorMatchingInlineSnapshot();
});

