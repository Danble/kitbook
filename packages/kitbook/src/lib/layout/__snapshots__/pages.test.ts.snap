// Vitest Snapshot v1

exports[`putPagesIntoFolders organizes pages into proper folders 1`] = `
{
  "depth": 0,
  "folders": [
    {
      "depth": 1,
      "folders": [
        {
          "depth": 2,
          "folders": [],
          "name": "my notes",
          "pages": [
            {
              "ext": "md",
              "name": "unocss",
              "path": "/src/docs/my-notes/0-unocss.md",
              "url": "/docs/my-notes/0-unocss",
            },
            {
              "ext": "md",
              "name": "deploy to vercel",
              "path": "/src/docs/my-notes/1-deploy-to-vercel.md",
              "url": "/docs/my-notes/1-deploy-to-vercel",
            },
          ],
          "url": "/docs/my-notes",
        },
      ],
      "name": "docs",
      "pages": [
        {
          "ext": "md",
          "name": "why kitbook",
          "path": "/src/docs/0-why-kitbook.md",
          "url": "/docs/0-why-kitbook",
        },
        {
          "ext": "md",
          "name": "get started",
          "path": "/src/docs/1-get-started.md",
          "url": "/docs/1-get-started",
        },
        {
          "ext": "md",
          "name": "index",
          "path": "/src/docs/index.md",
          "url": "/docs/index",
        },
      ],
      "url": "/docs",
    },
    {
      "depth": 1,
      "folders": [
        {
          "depth": 2,
          "folders": [],
          "name": "a",
          "pages": [
            {
              "ext": "svelte",
              "name": "C",
              "path": "/src/lib/a/C.svelte",
              "url": "/lib/a/C",
            },
            {
              "ext": "variants.ts",
              "name": "C",
              "path": "/src/lib/a/C.variants.ts",
              "url": "/lib/a/C",
            },
            {
              "ext": "svelte",
              "name": "D",
              "path": "/src/lib/a/D.svelte",
              "url": "/lib/a/D",
            },
            {
              "ext": "svx",
              "name": "D",
              "path": "/src/lib/a/D.svx",
              "url": "/lib/a/D",
            },
            {
              "ext": "variants.ts",
              "name": "D",
              "path": "/src/lib/a/D.variants.ts",
              "url": "/lib/a/D",
            },
          ],
          "url": "/lib/a",
        },
      ],
      "name": "lib",
      "pages": [
        {
          "ext": "svelte",
          "name": "A",
          "path": "/src/lib/A.svelte",
          "url": "/lib/A",
        },
        {
          "ext": "svelte",
          "name": "B",
          "path": "/src/lib/B.svelte",
          "url": "/lib/B",
        },
        {
          "ext": "svx",
          "name": "B",
          "path": "/src/lib/B.svx",
          "url": "/lib/B",
        },
        {
          "ext": "svx",
          "name": "E",
          "path": "/src/lib/E.svx",
          "url": "/lib/E",
        },
        {
          "ext": "foo.svelte",
          "name": "A",
          "path": "/src/lib/A.foo.svelte",
          "url": "/lib/A",
        },
      ],
      "url": "/lib",
    },
    {
      "depth": 1,
      "folders": [
        {
          "depth": 2,
          "folders": [],
          "name": "a",
          "pages": [
            {
              "ext": "svelte",
              "name": "+page",
              "path": "/src/routes/a/+page.svelte",
              "url": "/routes/a/+page",
            },
            {
              "ext": "svx",
              "name": "_page",
              "path": "/src/routes/a/_page.svx",
              "url": "/routes/a/_page",
            },
          ],
          "url": "/routes/a",
        },
        {
          "depth": 2,
          "folders": [],
          "name": "b",
          "pages": [
            {
              "ext": "svelte",
              "name": "+page",
              "path": "/src/routes/b/+page.svelte",
              "url": "/routes/b/+page",
            },
            {
              "ext": "variants.ts",
              "name": "_page",
              "path": "/src/routes/b/_page.variants.ts",
              "url": "/routes/b/_page",
            },
          ],
          "url": "/routes/b",
        },
        {
          "depth": 2,
          "folders": [],
          "name": "c",
          "pages": [
            {
              "ext": "svelte",
              "name": "+page",
              "path": "/src/routes/c/+page.svelte",
              "url": "/routes/c/+page",
            },
            {
              "ext": "svx",
              "name": "_page",
              "path": "/src/routes/c/_page.svx",
              "url": "/routes/c/_page",
            },
            {
              "ext": "variants.ts",
              "name": "_page",
              "path": "/src/routes/c/_page.variants.ts",
              "url": "/routes/c/_page",
            },
          ],
          "url": "/routes/c",
        },
      ],
      "name": "routes",
      "pages": [
        {
          "ext": "svelte",
          "name": "+page",
          "path": "/src/routes/+page.svelte",
          "url": "/routes/+page",
        },
      ],
      "url": "/routes",
    },
  ],
  "name": ".",
  "pages": [
    {
      "ext": "md",
      "name": "README",
      "path": "/README.md",
      "url": "/README",
    },
  ],
  "url": "/",
}
`;
